OpenSSH presented as OtherSSH in settings

gerhardol opened the pull request:
Part of #9112

## Proposed changes

Some SSH refactoring has changed the SSH setting to not always be stored as "" (stored in registry).
The OpenSSH setting is therefore always presented as other.

This is part of what described in #9112, but not the core problem.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

OpenSSH is not set, but the path is correct

![image](https://user-images.githubusercontent.com/6248932/117075300-dfd33e80-ad34-11eb-9b95-db6b995a94e6.png)

### After

![image](https://user-images.githubusercontent.com/6248932/117075584-5e2fe080-ad35-11eb-9a83-1585a9d07bf1.png)

## Test methodology <!-- How did you ensure quality? -->

manual

---

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented on GitCommands/SshPathLocator.cs:
```diff
@@ -68,5 +74,13 @@ public string Find(string gitBinDirectory)
         }
+
+        /// <summary>
+        /// Check if a path is a known OpenSSH executable
```
The implementation is quite different from what the doc states.
```suggestion
        /// Checks if OpenSSH is accessible via %PATH%, or comes from the git's bin folder.
```

gerhardol replied:
Will remove this method, add an out parameter to Find() instead to inform  if this is an OpenSSH - or we go the full way and set it to empty if matching OpenSSH in the Git bindir. (The last alternative may be OK, I was not sure when I did this initially.)

RussKie replied:
That seems like a better solution.

gerhardol commented:
I will change this to an `out` parameter in Find() instead.

RussKie commented on GitUI/CommandsDialogs/SettingsDialog/Pages/ChecklistSettingsPage.cs:
```diff
@@ -447,7 +447,8 @@ private bool CheckSSHSettings()
 
             var ssh = _sshPathLocator.Find(AppSettings.GitBinDir);
-            RenderSettingSet(SshConfig, SshConfig_Fix, string.IsNullOrEmpty(ssh) ? _opensshUsed.Text : string.Format(_unknownSshClient.Text, ssh));
+            var sshPath = _sshPathLocator.Find(AppSettings.GitBinDir);
```
`_sshPathLocator.Find` return a valid path or an empty string, if SSH is not found. Why do we need to detect it the second time?

gerhardol replied:
To find if this is OpenSSH or not.
This will be changed (see above), probably returning empty for OpenSSH.

RussKie commented:
> Some SSH refactoring has changed the SSH setting to not always be stored as "" (stored in registry).

I think we need to fix this. Are you able to provide more information?

gerhardol commented:
> > Some SSH refactoring has changed the SSH setting to not always be stored as "" (stored in registry).
> 
> I think we need to fix this. Are you able to provide more information?

If registry key (gitssh) is empty, the GIR_SSH environment variable is not set and Git will use the bundled ssh.exe.
If the configured path is empty, OpenSSH is used.

The refactoring will however try to set a value in the registry and within the application.
ssh works mostly fine anyway, a few annoyances:
* The setting always show "Other", never OpenSSH (the path is OK though)
* If you switch Git path, the ssh path is stuck on the old Git path

This was a fix for presentation primarily, to include in 3.5 and to simplify the more severe issue with SSH path being modified incorrectly when installing 3.5.

gerhardol commented on GitCommands/SshPathLocator.cs:
```diff
@@ -43,6 +44,11 @@ public string Find(string gitBinDirectory)
             return ssh ?? "";
         }
 
+        /// <summary>
+        /// Get ssh path from Git installation
+        /// </summary>
+        /// <param name="gitBinDirectory">Git installation directory</param>
+        /// <returns>Path to ssh.exe or null</returns>
```
will update

gerhardol commented:
Changed back to use "" for OpenSSH. As most original changes were reverted, I squashed on master.
The original commit for SSH Locator mentioned that the change was done to always set GIT_SSH env var, not properly done (variable unset when saving settings). A second commit corrects that. As the env var is set regardless, there is no change for the binary.

RussKie commented on GitCommands/Git/GitSshHelpers.cs:
```diff
@@ -26,19 +26,16 @@ bool DoArgumentsRequireSsh()
+            if (Strings.IsNullOrEmpty(path))
             {
-                Environment.SetEnvironmentVariable("GIT_SSH", path, EnvironmentVariableTarget.Process);
+                // GIT_SSH is not required by Git to use OpenSSH, but supposedly by Gerrit plugin
+                path = _sshPathLocatorInstance.GetSshFromGitDir(AppSettings.GitBinDir) ?? "";
```
🤔 If this is required by Gerrit, should this be in Gerrit plugin?

gerhardol replied:
Maybe
It does not seem like Gerrit uses GIT_SSH directly but relies on SshPathLocator.Find().
It could probably be OK to unset the variable if empty instead.
Pushing an update.

Gerrit maybe needs to implement SshPathLocator.GetSshFromGitDir() too, not sure how an empty path is handled. (There is code, not sure if that is working.)
The plugin imports GitUI so it need to be rebuilt on most GE changes anyway.

mstv reviewed (approved):
LGTM, have not run

RussKie commented:
Rebase and merge
